### ISSUE TITLE
Using VFL loss in Yolov8 but the dimensions of the anchor class and gt class not match.

### ISSUE THREAD

**Issue by ricky-696** [ultralytics/ultralytics#3604]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I want to use VFL loss because my dataset's label is imbalance, I see the issue in #1448 

I think this issue occurs because the second dimension of `pred_scores` and `target_scores` represents the number of anchor boxes, while the second dimension of `gt_labels` represents the number of ground truth bounding boxes. 

https://github.com/ultralytics/ultralytics/blob/8940a27bdb26895f09a1554514a9a46312aa89c3/ultralytics/yolo/utils/loss.py#L131

I want to know how to extract the number of class of the anchor boxes from the model's output in this function.

This my error, same as #1448: 
```
Exception has occurred: RuntimeError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
The size of tensor a (33600) must match the size of tensor b (156) at non-singleton dimension 1
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/site-packages/ultralytics/yolo/utils/loss.py", line 24, in forward
    weight = alpha * pred_score.sigmoid().pow(gamma) * (1 - label) + gt_score * label
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/site-packages/ultralytics/yolo/utils/loss.py", line 165, in __call__
    loss[1] = self.varifocal_loss(pred_scores, target_scores, gt_labels) / target_scores_sum  # VFL way
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 213, in loss
    return self.criterion(self.predict(batch['img']) if preds is None else preds, batch)
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 44, in forward
    return self.loss(x, *args, **kwargs)
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/site-packages/ultralytics/yolo/engine/trainer.py", line 330, in _do_train
    self.loss, self.loss_items = self.model(batch)
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/site-packages/ultralytics/yolo/engine/trainer.py", line 192, in train
    self._do_train(world_size)
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/site-packages/ultralytics/yolo/engine/model.py", line 371, in train
    self.trainer.train()
  File "/home/Ricky/0_Project/ROADpp_challenge_ICCV2023/Track1/train_YOLOv8.py", line 9, in <module>
    results = model.train(
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/Ricky/.conda/envs/ROAD/lib/python3.10/runpy.py", line 196, in _run_module_as_main (Current frame)
    return _run_code(code, main_globals, None,
RuntimeError: The size of tensor a (33600) must match the size of tensor b (156) at non-singleton dimension 1
```

I think 33600 is the number of anchor boxes, and 156 is the number of gt boxes, thanks for help!

**Comment by glenn-jocher** [comment]
@ricky-696 hi there,

It seems that you are experiencing an issue with the dimensions of the anchor class and ground truth class not matching when using the VFL (Varifocal) loss in YOLOv8.

In YOLOv8, the second dimension of `pred_scores` and `target_scores` represents the number of anchor boxes, while the second dimension of `gt_labels` represents the number of ground truth bounding boxes. The number of classes for the anchor boxes can be extracted from the model's output in the `varifocal_loss` function in the `loss.py` file.

The error you provided indicates that the size of tensor `a` (33600) must match the size of tensor `b` (156) at non-singleton dimension 1. This suggests that there is a discrepancy between the number of anchor boxes (33600) and the number of ground truth boxes (156). 

To resolve this issue, you will need to ensure that the number of anchor boxes aligns with the number of ground truth boxes in your dataset. This could involve adjusting the dimensions of the anchor boxes or modifying the ground truth labels.

I hope this helps! Let me know if you have any further questions.

**Comment by ricky-696** [comment]
Hello @glenn-jocher, thanks for help!
I think the `varifocal_loss` function you mean is here: 
https://github.com/ultralytics/ultralytics/blob/8940a27bdb26895f09a1554514a9a46312aa89c3/ultralytics/yolo/utils/loss.py#L15

My question is the number of ground truth (gt) boxes will always be different from the number of anchor boxes. In this function, `gt_score * label` will definitely cause an error. If I want to use this function, do I need to know how Yolov8 selects anchor boxes as the final result boxes or how the loss is actually calculated?

I think I need to understand how VFL loss is calculated.

**Comment by sunmooncode** [comment]
@ricky-696  I have encountered the same problem, Have you successfully implemented VarifocalLoss in yolov8?

**Comment by glenn-jocher** [comment]
@sunmooncode hi there! The process of successfully implementing VarifocalLoss in YOLOv8 involves navigating and modifying certain components of the YOLOv8 model.

The issue that you have encountered where the dimensions of the ground truth and anchor class do not match can be attributed to the function varifocal_loss. The VarifocalLoss (VFL) is designed to handle datasets with imbalanced class labels. However, in applying VFL to YOLOv8, the second dimension of the predicted scores and the target scores corresponds to the number of anchor boxes whereas the second dimension for the ground truth labels reflects the number of ground truth bounding boxes. This discrepancy in the dimensions can cause an error as you have observed.

To resolve this dimension mismatch, you would need to adjust the number of anchor boxes to be equal to the number of ground truth boxes. This adjustment might involve modifying the dimensions of the anchor boxes or adjusting the ground truth labels to match the number of anchor boxes.

As you continue to face challenges, knowing the workings of YOLOv8's anchor box selection mechanism and understanding the loss calculation process would be beneficial. Additionally, gaining an intuitive understanding of how VFL loss is computed will also help you to successfully implement VarifocalLoss in YOLOv8.

**Comment by ricky-696** [comment]
Hi @sunmooncode 
I ultimately chose to incorporate the YOLOv7's `QFocalLoss` into YOLOv8 in an attempt to address my label imbalance issue: 

`loss[1] = self.QFocalLoss(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum # FLv7`

and the `QFocalLoss` class in here:
https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/data/inference_dataset.py#L593

Unfortunately, adding the FocalLoss-based loss function did not lead to an improvement in the model's mAP.

**Comment by sunmooncode** [comment]
@ricky-696 Thank you for your work!
Since my job involves small object detection, I am interested in using VarifocalLoss as it seems to improve the performance of small object detection.

**Comment by glenn-jocher** [comment]
@ricky-696, thank you for your interest in VarifocalLoss! You're right, VarifocalLoss can greatly improve the performance of small object detection by addressing the problem of foreground and background class imbalance.

In YOLOv8, implementing VarifocalLoss would involve a few modifications. Firstly, VarifocalLoss requires the prediction score, which is essentially the class probability that the network outputs for each bounding box. This information would be needed to adjust the class imbalance.

Secondly, because YOLOv8 uses a multiple bounding box prediction strategy, you would need to ensure consistent dimensions between the ground truth labels and the predicted scores for each bounding box. In simpler terms, every predicted bounding box must have a corresponding target score. This may involve modifying the dimensions of the respective tensors.

The critical task would be to balance the weight of the positive and negative classes during the loss calculation, which is achieved by VarifocalLoss, but requires very careful adjustments and proper normalization.

Be aware that the integration of VarifocalLoss into YOLOv8 may necessitate additional modifications. Also, while VarifocalLoss has proven to be effective in some cases, its performance may vary depending on the specific contexts and datasets.

Good luck with your modifications! Optimizing the object detection performance for smaller objects is a challenging but certainly achievable task. Let us know how it goes.

**Comment by Hxiaowu666** [comment]
May I ask what “label” in VarifocalLoss specifically refers to? I don't see any relevant information in the literature theory and code of vairfocal loss

**Comment by glenn-jocher** [comment]
@Hxiaowu666 hi there,

In the context of VarifocalLoss, the term "label" typically refers to the ground truth label associated with each bounding box during the training process. Labels are usually encoded as one-hot vectors, where each element corresponds to a specific class in your dataset.

In a broader context, when an image is passed through the model, it predicts a set of bounding boxes along with respective confidence scores and class probabilities. The aim of VarifocalLoss is to compute the difference between the predicted and the actual ground truth (i.e., the "label"). 

In simple terms, if you have a dog in the image, the label for this instance would be 'dog', and if the model predicts this instance as 'cat' with a high confidence score, VarifocalLoss will consider this as a hard example and assign a higher weight to its loss.

Understanding how VarifocalLoss operates with these "labels" helps the model focus more on harder positives and easier negatives, addressing class imbalance during the training process.

Hope this clarifies your question.

**Comment by JiangnanCai** [comment]
> @sunmooncode hi there! The process of successfully implementing VarifocalLoss in YOLOv8 involves navigating and modifying certain components of the YOLOv8 model.
> 
> The issue that you have encountered where the dimensions of the ground truth and anchor class do not match can be attributed to the function varifocal_loss. The VarifocalLoss (VFL) is designed to handle datasets with imbalanced class labels. However, in applying VFL to YOLOv8, the second dimension of the predicted scores and the target scores corresponds to the number of anchor boxes whereas the second dimension for the ground truth labels reflects the number of ground truth bounding boxes. This discrepancy in the dimensions can cause an error as you have observed.
> 
> To resolve this dimension mismatch, you would need to adjust the number of anchor boxes to be equal to the number of ground truth boxes. This adjustment might involve modifying the dimensions of the anchor boxes or adjusting the ground truth labels to match the number of anchor boxes.
> 
> As you continue to face challenges, knowing the workings of YOLOv8's anchor box selection mechanism and understanding the loss calculation process would be beneficial. Additionally, gaining an intuitive understanding of how VFL loss is computed will also help you to successfully implement VarifocalLoss in YOLOv8.

hi @glenn-jocher , I am confused about the meaning of third dimension of 'target_scores', is it correspondes to the class number of dataset? if the score is not 0, what does it mean?

**Comment by JiangnanCai** [comment]
@glenn-jocher thanks for your explanation about 'target_score'. 

to understand the anchor box selection mechanism, I read the code in /ultralytics/utils/tal.py. 

line 136：`align_metric, overlaps = self.get_box_metrics(pd_scores, pd_bboxes, gt_labels, gt_bboxes, mask_in_gts * mask_gt)`

the 'mask_in_gts' variable is generated from the function [select_candidates_in_gts](https://github.com/ultralytics/ultralytics/blob/577d066fb201a62694ed1e9d53f715e27db0266e/ultralytics/utils/tal.py#L12-L28C1).

Based on my observations, it appears that all ground truth bounding boxes (gt_bboxes) with zero values in their third dimension, which represents the location of the gt_bboxes, are set to a negative value after the operation bbox_deltas.amin(3), effectively filtering them out (setting them to False). This implies that the mask 'mask_in_gts' already encodes information about the existence of ground truth boxes.

Additionally, 'mask_gt' is obtained from [line 176 in /ultralytics/utils/loss.py](https://github.com/ultralytics/ultralytics/blob/577d066fb201a62694ed1e9d53f715e27db0266e/ultralytics/utils/loss.py#L176) as mask_gt = gt_bboxes.sum(2, keepdim=True).gt_(0), and it also indicates the existence of ground truth boxes.

I am curious why does the code use `mask_in_gts * mask_gt` instead of just `mask_in_gts`?

**Comment by glenn-jocher** [comment]
Hi @JiangnanCai,

Great question. The purpose of using `mask_in_gts * mask_gt` instead of just `mask_in_gts` is to ensure that both conditions are satisfied simultaneously, thereby refining the selection of candidate proposal boxes to match with ground truth boxes during training.

"mask_in_gts" is used to filter out invalid ground truth bounding boxes, effectively focusing on those ground truth boxes that are supposed to be detected. This is because only valid ground truth bounding boxes carry significant information for the model to learn.

On the other hand, "mask_gt" serves as an additional check to ensure that the ground truth bounding box indeed exists, i.e., it is not just a placeholder or padding used for maintaining consistent tensor sizes across batches. Those boxes would contribute zero to the overall loss and are not supposed to impact the learning of the model.

Therefore, the multiplication (`mask_in_gts * mask_gt`) is a way to apply both filtering conditions. It effectively zeroes out any instances where either the ground truth box is invalid (as per 'mask_in_gts') or no actual ground truth box exists (as per 'mask_gt').

In essence, combining "mask_in_gts" and "mask_gt" ensures a more robust and reliable selection of anchor boxes that contribute to the actual learning and thus improves the performance of the model. Hopefully this explanation helps!

**Comment by JiangnanCai** [comment]
Hi @glenn-jocher,

Thanks a lot for your previous explanations. 

now I am trying to convert current `target_labels`, which is provided by tal, to a one-hot tensor to be used in vfl.

current 'target_labels' is in shape (16, 8400). (batchsize 16).

```python
tensor([[11, 11, 11,  ..., 11, 11, 11],
        [11, 11, 11,  ..., 11, 11, 11],
        [ 0,  0,  0,  ...,  0,  0,  0],
        ...,
        [ 9,  9,  9,  ...,  9,  9,  9],
        [ 8,  8,  8,  ...,  8,  8,  8],
        [ 2,  2,  2,  ...,  2,  2,  2]], device='cuda:0')
``` 

so i use following codes to convert it to a one-hot tensor with same shape as `pred_scores` and `targets_scores`, which is (16, 8400, 13).

```python
target_labels = target_labels.unsqueeze(-1).expand(-1, -1, self.nc)  # self.nc: class num
one_hot = torch.zeros(target_labels.size(), device=self.device)
one_hot.scatter_(-1, target_labels, 1)
```

after the conversion, the `target_labels` is in shape (16, 8400, 13). In the third dimension, all are set to zero and only index==label would be set to one. 

is that the final 'target_labels' that we need in the vfl?

**Comment by JiangnanCai** [comment]
Hi @glenn-jocher

I start to train with varifocal loss last weekend and get 2 questions about it.

1. the varifocal loss is extremly small (about 10-e5) compare to box loss and dfl loss. 

2. the implement of varifocal loss in code

the official function of varifocal loss is:
$$\text{VFL}(p, q) = \left\{\begin{matrix}
  -q(q \log(p) + (1-q)log(1-p)) & q > 0 \\ 
 -\alpha p^\gamma \log(1-p) & q=0
\end{matrix}\right.$$

and with the help of bce loss and `target_labels`, it can be simplified to following form:

$$\text{VFL}(p, q) = -\alpha p^{\gamma}\cdot (1- \text{target_labels}) \cdot \text{BCE}(p, q) + q \cdot \text{target_labels} \cdot \text{BCE}(p, q)  $$

where $q$ indicates `gt_score` and $p$ indicates `pred_scores`.

It seems a little bit diffierence between the code implementation and upper expression.  The code implementation missing a minus sign before $\alpha$.

Am I decude the expression incorrectly or there are some special reasons for that.

**Comment by glenn-jocher** [comment]
@JiangnanCai hi there,

Let's address your concerns one by one:

1. Varifocal Loss (VFL) Being Extremely Small:
The VFL can indeed be significantly smaller than the box loss and distribution focal loss (DFL). This may happen due to the multiplicative term that applies a heavy down-weighting for easy negative examples, which are often the dominant type of examples. This behavior is expected since VFL focuses more on hard examples (i.e., examples that the model currently gets wrong) and assigns smaller loss contributions to those it already handles well.

2. Implementation of Varifocal Loss:

The equation you've referred to from the official definition of Varifocal Loss is the key to understanding its implementation. In mathematical notation, the code should implement the function as:

$$
\text{VFL}(p, q) = -\alpha \cdot p^{\gamma} \cdot (1 - q) \cdot \log(p) + q \cdot \log(1 - p)
$$

Here you're right to observe the discrepancy related to the minus sign signifying that the first part should contribute a positive loss for incorrect predictions with high predicted scores. As for the actual code implementation, please carefully review whether the code you're referencing has an implicit handling of a negative sign elsewhere, such as within the BCE implementation itself. Sometimes, deep learning frameworks such as PyTorch apply some operations in a more complex manner that can include implicit negations to fit their optimization paradigms.

If the minus sign is genuinely missing and not compensated for elsewhere, it might mean that the loss is being minimized in the opposite direction than intended, but this would typically result in model divergence rather than convergence.

To align with the proper functionality of Varifocal Loss, do make sure that the implementation punishes for wrong predictions and high confidence (where the label is zero and the prediction is high), which should be reflected as a positive loss, requiring the minus sign.

Without seeing the exact code implementation, I suggest closely comparing against the official definition and confirming that each element is implemented as expected, including the negative sign in the correct place to ensure that the loss behaves as intended.

Keep in mind that the AI and ML community relies on collective expertise and comprehensive testing, so if you do encounter what seems to be a genuine issue with widely used code, reporting this to the repository helps everyone. Your diligence is appreciated, and I'm here to help clarify any further concerns.

**Comment by hellowqc** [comment]
> Hi @glenn-jocher,
> 
> Thanks a lot for your previous explanations.
> 
> now I am trying to convert current `target_labels`, which is provided by tal, to a one-hot tensor to be used in vfl.
> 
> current 'target_labels' is in shape (16, 8400). (batchsize 16).
> 
> ```python
> tensor([[11, 11, 11,  ..., 11, 11, 11],
>         [11, 11, 11,  ..., 11, 11, 11],
>         [ 0,  0,  0,  ...,  0,  0,  0],
>         ...,
>         [ 9,  9,  9,  ...,  9,  9,  9],
>         [ 8,  8,  8,  ...,  8,  8,  8],
>         [ 2,  2,  2,  ...,  2,  2,  2]], device='cuda:0')
> ```
> 
> so i use following codes to convert it to a one-hot tensor with same shape as `pred_scores` and `targets_scores`, which is (16, 8400, 13).
> 
> ```python
> target_labels = target_labels.unsqueeze(-1).expand(-1, -1, self.nc)  # self.nc: class num
> one_hot = torch.zeros(target_labels.size(), device=self.device)
> one_hot.scatter_(-1, target_labels, 1)
> ```
> 
> after the conversion, the `target_labels` is in shape (16, 8400, 13). In the third dimension, all are set to zero and only index==label would be set to one.
> 
> is that the final 'target_labels' that we need in the vfl?

Hello, sorry to bother you. I would like to ask the target_ How did I obtain the labels? I didn't see this variable in loss. py

**Comment by hellowqc** [comment]
> ```python
> ```python
> one_hot.scatter_(-1, target_labels, 1)
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```

Hello, I think there was an error in your conversion. Have you resolved it?

**Comment by glenn-jocher** [comment]
@hellowqc hello,

It seems there might be a misunderstanding. The code snippet you've posted is part of a process to convert class indices into a one-hot encoded format, which is a common step when preparing targets for certain types of loss functions. If you're encountering an issue with this conversion, could you please provide more details about the error or the problem you're facing? This will help me assist you more effectively.

**Comment by hellowqc** [comment]
> @hellowqc hello,
> 
> It seems there might be a misunderstanding. The code snippet you've posted is part of a process to convert class indices into a one-hot encoded format, which is a common step when preparing targets for certain types of loss functions. If you're encountering an issue with this conversion, could you please provide more details about the error or the problem you're facing? This will help me assist you more effectively.

target_labels = target_labels.unsqueeze(-1).expand(-1, -1, self.nc)  # self.nc: class num
        one_hot = torch.zeros(target_labels.size(), device=self.device)
        one_hot.scatter_(-1, target_labels, 1)  
This is the specific conversion process. Is this conversion correct?

**Comment by glenn-jocher** [comment]
@hellowqc,

The conversion process you've described for creating a one-hot encoded tensor from `target_labels` appears to be correct. This process will result in a tensor where each class index in `target_labels` is represented as a one-hot vector along the last dimension. If `self.nc` is the number of classes, then the resulting `one_hot` tensor should have the shape `(batch_size, number_of_anchors, number_of_classes)`, with each slice along the last dimension being a one-hot vector indicating the class of the corresponding anchor. If you encounter any issues with this conversion, please provide the error message or describe the problem for further assistance.

**Comment by hellowqc** [comment]
> @hellowqc,
> 
> The conversion process you've described for creating a one-hot encoded tensor from `target_labels` appears to be correct. This process will result in a tensor where each class index in `target_labels` is represented as a one-hot vector along the last dimension. If `self.nc` is the number of classes, then the resulting `one_hot` tensor should have the shape `(batch_size, number_of_anchors, number_of_classes)`, with each slice along the last dimension being a one-hot vector indicating the class of the corresponding anchor. If you encounter any issues with this conversion, please provide the error message or describe the problem for further assistance.

Thank you for your reply

I tried doing this, but during training, the classification loss would decrease to a very small value, and the final mAP50 would also be very small compared to using BCE loss.

I was wondering if I made a mistake in converting?

**Comment by glenn-jocher** [comment]
@hellowqc,

The one-hot conversion process you've implemented is standard and should not inherently cause the issues you're experiencing. A very small classification loss and lower mAP50 could be indicative of other factors at play, such as an imbalance in the dataset, learning rate issues, or the model not learning the features well enough to distinguish between classes. It's also possible that the loss function or its integration with the training loop may need to be reviewed to ensure it's being applied correctly. If the model was performing better with BCE loss, it might be worth comparing the behavior of both loss functions during training to identify any discrepancies.

**Comment by hellowqc** [comment]
> @hellowqc,
> 
> The one-hot conversion process you've implemented is standard and should not inherently cause the issues you're experiencing. A very small classification loss and lower mAP50 could be indicative of other factors at play, such as an imbalance in the dataset, learning rate issues, or the model not learning the features well enough to distinguish between classes. It's also possible that the loss function or its integration with the training loop may need to be reviewed to ensure it's being applied correctly. If the model was performing better with BCE loss, it might be worth comparing the behavior of both loss functions during training to identify any discrepancies.

But when I use cross entropy loss as the classification loss, mAP50 can reach 90. After replacing it with focus loss using the above conversion method, mAP50 is only about 50. The difference between the two is too great, so is there a conversion error?

**Comment by glenn-jocher** [comment]
@hellowqc,

If you're seeing a significant drop in mAP50 when switching from BCE to Focal Loss with the one-hot encoding method you've described, it suggests that the issue may not be with the conversion itself but rather with how the Focal Loss is being applied or integrated with the rest of the model. Focal Loss has hyperparameters, such as the focusing parameter gamma and the alpha weighting, which need to be tuned for your specific dataset and task. It's also important to ensure that the loss is being calculated and backpropagated correctly. Consider revisiting the Focal Loss implementation and its parameters, and compare it closely with the BCE setup to find any potential discrepancies.

**Comment by saikrishna0203** [comment]
target_labels = target_labels.unsqueeze(-1).expand(-1, -1, self.nc)  # self.nc: class num
one_hot = torch.zeros(target_labels.size(), device=self.device)
one_hot.scatter_(-1, target_labels, 1)  whee i have to add these lines for getting varifocal loss @glenn-jocher  pls give me steps to compute,presently i initializedvarifocal loss and uncommented the 185 line and commented 186 line ,now what i have to do and where i have to keep those lines ,target_labeels means those txt files which have class id and coordinates?

**Comment by saikrishna0203** [comment]
@JiangnanCai did u get how to  get varifocal loss?
